### PR TITLE
Introducing User Resource + removing User.find callsites

### PR DESCRIPTION
**Pull request by albandum**
## Description

This PR introduces a new UserResource class to centralize user-related operations and replace direct User model interactions. It removes the renderUserType function and updates all callsites to use UserResource methods instead. The change aims to improve code organization, type safety, and consistency in user-related operations across the codebase.

This PR:
- Removes `renderUserType` function
- Introduces UserResource with the following methods:
```
- makeNew
- fetchAllByModelIds
- fetchAllByUsername
- fetchAllByEmail
- fetchByExternalId
- fetchByAuth0Sub
- fetchByEmail
- fetchByProvider
- getWorkspaceFirstAdmin
- update
- delete
- toJSON
```

- Replaces all direct calls to User.findOne() with calling the resource.
- There's only a few complex User.find(x) left about documenttracker. cc @fontanierh do we want to migrate that to the resource as well?

## Risk
This change impacts user fetching and manipulation throughout the application. There's a risk of breaking user-related functionality if any use cases were missed during the refactoring. Thorough testing of user authentication, profile management, and any features involving user data is crucial.

## Deploy Plan

Tested on edge
- Registration
- Inviting new user to workspace
- Inviting existing dust user to workspace
- Revoking user
- auth0 login (google)
- Switching workspaces
- Removing user from poke
- Conversations reactions
- workspace deletion (scrub)
- OKTA

To keep under surveillance
- CustomerIO behavior

### CODE REVIEW

File: front/lib/resources/user_resource.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      where: {
+        username,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchByExternalId(
+    sid: string,
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:35]: `userId`

We generally name the sId these way through interfaces

File: front/lib/resources/user_resource.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    transaction?: Transaction
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:36]: We generally don't want transactions in the interface of Resources cc @flvndvd

Comment (spolu) [reply front/lib/resources/user_resource.ts:36]: I presume there's no way around all these for now... They will be removable as we have higher and higher level resources

File: front/lib/resources/user_resource.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        transaction,
+      });
+
+      return new Ok(undefined);
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  toJSON(): User {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:140]: Nope toJSON should return a `UserType` not `User` which is not a serializable thing

File: front/lib/resources/user_resource.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    return this.model.build(this);
+  }
+
+  toUserType(): UserType {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:144]: To be removed once toJSON is updated

**Review by spolu — COMMENTED** [review]
First batch of comments post reviewing the resource itself

File: front/admin/cli.ts [hunk 5]
```diff
@@ -218,7 +208,7 @@ const user = async (command: string, args: parseArgs.ParsedArgs) => {
       console.log(`  email: ${u.email}`);
 
       const memberships = await MembershipResource.getLatestMemberships({
-        users: [renderUserType(u)],
+        users: [u.toUserType()],
```
Comment (spolu) [inline front/admin/cli.ts:211]: toJSON

File: front/lib/resources/user_resource.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        auth0Sub: sub,
+      },
+      transaction,
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByEmail(
+    email: string,
+    transaction?: Transaction
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:62]: This one is not needed as per reviewing the codebase

File: front/lib/resources/user_resource.ts (hunk 2 above)
Comment (spolu) [reply front/lib/resources/user_resource.ts:36]: Actually this one can be removed from reviewing the rest of the codebase

File: front/lib/resources/user_resource.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        sId: sid,
+      },
+      transaction,
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByAuth0Sub(
+    sub: string,
+    transaction?: Transaction
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:49]: This transaction can be removed

File: front/lib/iam/users.ts [hunk 8]
```diff
@@ -18,29 +19,23 @@ async function fetchUserWithLegacyProvider(
...
+    providerId.toString()
+  );
 
   // If a legacy user is found, attach the Auth0 user ID (sub) to the existing user account.
   if (user) {
     await user.update({ auth0Sub: sub });
   }
 
-  return user;
+  return user?.toJSON();
```
Comment (spolu) [inline front/lib/iam/users.ts:32]: The resource should be returned here

File: front/lib/iam/users.ts [hunk 9]
```diff
@@ -18,29 +19,23 @@ async function fetchUserWithLegacyProvider(
...
 async function fetchUserWithAuth0Sub(sub: string) {
-  const userWithAuth0 = await User.findOne({
-    where: {
-      auth0Sub: sub,
-    },
-  });
+  const userWithAuth0 = await UserResource.fetchByAuth0Sub(sub);
 
-  return userWithAuth0;
+  return userWithAuth0?.toJSON();
```
Comment (spolu) [inline front/lib/iam/users.ts:38]: The resource should be returned here

File: front/lib/workspace.ts [hunk 10]
```diff
@@ -28,17 +27,6 @@ export function renderLightWorkspaceType({
...
-          role: "admin",
-          workspaceId: workspace.id,
-        },
-        required: true,
-      },
-    ],
-    order: [["createdAt", "ASC"]],
-  });
+  const userRes = await UserResource.getWorkspaceFirstAdmin(workspace.id);
+  return userRes?.toJSON();
```
Comment (spolu) [inline front/lib/workspace.ts:31]: The resource should be returned here

File: front/poke/temporal/activities.ts [hunk 11]
```diff
@@ -512,17 +513,11 @@ export async function deleteMembersActivity({
...
-        },
-        transaction: t,
-      });
-
+      const user = await UserResource.fetchByModelId(membership.userId, t);
       if (user) {
         const membershipsOfUser = await MembershipResource.getLatestMemberships(
           {
-            users: [renderUserType(user)],
+            users: [user.toUserType()],
```
Comment (spolu) [inline front/poke/temporal/activities.ts:520]: toJSON

File: front/poke/temporal/activities.ts [hunk 12]
```diff
@@ -539,9 +534,9 @@ export async function deleteMembersActivity({
             `[Workspace delete] Deleting Membership ${membership.id} and user ${user.id}`
           );
           // Delete the user's files
-          await FileResource.deleteAllForUser(renderUserType(user), t);
+          await FileResource.deleteAllForUser(user.toUserType(), t);
```
Comment (spolu) [inline front/poke/temporal/activities.ts:537]: toJSON

File: front/lib/resources/user_resource.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      transaction,
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByProvider(
+    provider: UserProviderType,
+    providerId: string,
+    transaction?: Transaction
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:77]: can be removed

File: front/lib/resources/user_resource.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+      returning: true,
+    });
+
+    // Update the resource to reflect the new status.
+    Object.assign(this, affectedRows[0].get());
+  }
+
+  async delete(
+    auth: Authenticator,
+    transaction?: Transaction
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:124]: Can be removed?

File: front/lib/iam/users.ts [hunk 15]
```diff
@@ -95,46 +89,46 @@ export async function maybeUpdateFromExternalUser(
...
         );
-        user.firstName = firstName;
-        user.lastName = lastName;
+        newUser.firstName = firstName;
+        newUser.lastName = lastName;
       }
     }
 
-    await user.save();
+    await user.update({ ...newUser, createdAt: user.createdAt });
```
Comment (spolu) [inline front/lib/iam/users.ts:120]: Let's make it a specific method user.upateInfo maybe? So that we don't need that newUser weirdness

File: front/lib/resources/user_resource.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+      | "providerId"
+      | "imageUrl"
+    > &
+      Partial<Pick<Attributes<User>, "providerId" | "imageUrl">>
+  ): Promise<UserResource> {
+    const user = await User.create(blob);
+    return new this(User, user.get());
+  }
+
+  static async fetchAllByModelIds(ids: number[]): Promise<UserResource[]> {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:38]: Let's call it fetchByModelIds

Comment (spolu) [inline front/lib/resources/user_resource.ts:38]: Type it ModelId[]

File: front/lib/resources/user_resource.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    const users = await User.findAll({
+      where: {
+        id: ids,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchAllByUsername(username: string): Promise<UserResource[]> {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:48]: fetchByUserName as I believe we don't use the "All" vocabulary in other resources. Mostly for consistency of naming

File: front/lib/resources/user_resource.ts [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+    const users = await User.findAll({
+      where: {
+        username,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static fetchAllByEmail(email: string): Promise<UserResource[]> {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:58]: Same here?

Comment (spolu) [reply front/lib/resources/user_resource.ts:58]: Check with @flvndvd on preferred naming

Comment (spolu) [reply front/lib/resources/user_resource.ts:58]: Though we also have a fetchByEmail... Do we need that fetchAll ?

File: front/lib/resources/user_resource.ts [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+          required: true,
+        },
+      ],
+      order: [["createdAt", "ASC"]],
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  async update(blob: Partial<Attributes<User>>): Promise<void> {
```
Comment (spolu) [inline front/lib/resources/user_resource.ts:128]: Ideally we get rid of this one for more specific update methods?

File: front/migrations/20240423_backfill_customerio.ts [hunk 20]
```diff
@@ -31,7 +30,14 @@ const backfillCustomerIo = async (execute: boolean) => {
             }),
             // NOTE: this is unrelated to customerio, but leveraging this backfill
             // to also identify all users on Amplitude.
-            AmplitudeServerSideTracking._identifyUser({ user: u }),
+            AmplitudeServerSideTracking._identifyUser({
+              user: {
+                ...u.toJSON(),
+                fullName: `${u.firstName} ${u.lastName}`,
```
Comment (spolu) [inline front/migrations/20240423_backfill_customerio.ts:36]: Why is this needed?

File: front/pages/api/login.ts [hunk 21]
```diff
@@ -444,7 +444,7 @@ export async function createAndLogMembership({
   workspace,
   role,
 }: {
-  user: UserType;
+  user: UserResource;
```
Comment (spolu) [inline front/pages/api/login.ts:447]: cc @flvndvd do we want resources in lib/API signatures? I think no right?

File: front/pages/api/w/[wId]/members/[uId]/index.ts [hunk 22]
```diff
@@ -45,8 +45,8 @@ async function handler(
     });
   }
 
-  const user = await getUserForWorkspace(auth, { userId });
-  if (!user) {
+  const userRes = await getUserForWorkspace(auth, { userId });
```
Comment (spolu) [inline front/pages/api/w/[wId]/members/[uId]/index.ts:48]: Why call it userRes?

File: front/pages/no-workspace.tsx [hunk 23]
```diff
@@ -44,8 +45,18 @@ async function fetchRevokedWorkspace(
 ): Promise<Workspace | null> {
   // TODO(@fontanierh): this doesn't look very solid as it will start to behave
   // weirdly if a user has multiple revoked memberships.
+  const userRes = await UserResource.fetchByModelId(user.id);
+
+  if (!userRes) {
```
Comment (spolu) [inline front/pages/no-workspace.tsx:50]: This feels wrong. Do we want to keep user type as signature or do we want the auth to return a resource? cc @flvndvd

**Review by spolu — COMMENTED** [review]
Left some questions to discuss with @flvndvd but this  looks great!!

File: front/lib/api/user.ts [hunk 24]
```diff
@@ -2,61 +2,43 @@ import type { UserMetadataType, UserType } from "@dust-tt/types";
 
 import type { Authenticator } from "@app/lib/auth";
 import { User, UserMetadata } from "@app/lib/models/user";
+import { UserResource } from "@app/lib/resources/user_resource";
 
 import { MembershipResource } from "../resources/membership_resource";
 
-export function renderUserType(user: User): UserType {
```
Comment (flvndvd) [inline front/lib/api/user.ts:8]: 😘

File: front/lib/api/user.ts [hunk 25]
```diff
@@ -2,61 +2,43 @@ import type { UserMetadataType, UserType } from "@dust-tt/types";
...
   if (!membership) {
     return null;
   }
 
-  return renderUserType(user);
+  return user;
 }
 
-export async function deleteUser(user: UserType): Promise<void> {
+export async function deleteUser(user: UserResource): Promise<void> {
```
Comment (flvndvd) [inline front/lib/api/user.ts:41]: This should live in the `UserResource` now.

File: front/lib/iam/users.ts (hunk 15 above)
Comment (flvndvd) [reply front/lib/iam/users.ts:120]: 💯 ☝️

File: front/lib/iam/users.ts [hunk 26]
```diff
@@ -95,46 +89,46 @@ export async function maybeUpdateFromExternalUser(
...
 export async function createOrUpdateUser(
   session: SessionWithUser
-): Promise<{ user: UserType; created: boolean }> {
+): Promise<{ user: UserResource; created: boolean }> {
   const { user: externalUser } = session;
 
   const user = await fetchUserFromSession(session);
 
   if (user) {
+    const newUser = {} as Partial<UserType>;
```
Comment (flvndvd) [inline front/lib/iam/users.ts:98]: `newUser` feels weird here.

File: front/lib/resources/user_resource.ts [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static fetchAllByEmail(email: string): Promise<UserResource[]> {
+    return User.findAll({
+      where: {
+        email,
+      },
+    }).then((users) => users.map((user) => new UserResource(User, user.get())));
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:63]: We should not use "thenable", to keep consistency throughout the codebase.

File: front/lib/resources/user_resource.ts (hunk 18 above)
Comment (flvndvd) [reply front/lib/resources/user_resource.ts:58]: Convention should be:
- `fetch*` returns one entity.
- `list*` returns many entity.

File: front/lib/resources/user_resource.ts [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  static async fetchByExternalId(userId: string): Promise<UserResource | null> {
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:66]: This one should be `fetchById`.

File: front/lib/resources/user_resource.ts (hunk 19 above)
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:128]: We should not have a generic method like this, IIRC there is only one place were we update the users, it's in the login flow, let have a dedicated function here.

File: front/lib/resources/user_resource.ts [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+import { BaseResource } from "@app/lib/resources/base_resource";
+import { MembershipModel } from "@app/lib/resources/storage/models/membership";
+import type { ReadonlyAttributesType } from "@app/lib/resources/storage/types";
+
+// Attributes are marked as read-only to reflect the stateless nature of our Resource.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unsafe-declaration-merging
+export interface UserResource extends ReadonlyAttributesType<User> {}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class UserResource extends BaseResource<User> {
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:15]: Interesting to see all the weird ways we have to fetch a user 🤯.

File: front/lib/tracking/customerio/server.ts [hunk 30]
```diff
@@ -264,7 +261,7 @@ export class CustomerioServerSideTracking {
         email: user.email,
         first_name: user.firstName,
         last_name: user.lastName,
-        created_at: Math.floor(user.createdAt / 1000),
+        created_at: Math.floor(user.createdAt.getTime() / 1000),
```
Comment (flvndvd) [inline front/lib/tracking/customerio/server.ts:264]: IMHO we should have a toCustomerIo function on the `UserResource`.

File: front/lib/resources/user_resource.ts [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+    return {
+      sId: this.sId,
+      id: this.id,
+      createdAt: this.createdAt.getTime(),
+      provider: this.provider,
+      username: this.username,
+      email: this.email,
+      firstName: this.firstName,
+      lastName: this.lastName,
+      fullName: this.firstName + (this.lastName ? ` ${this.lastName}` : ""),
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:163]: Can we extract this into a getter so we can use it in different places?

File: front/migrations/20240423_backfill_customerio.ts [hunk 32]
```diff
@@ -31,7 +30,14 @@ const backfillCustomerIo = async (execute: boolean) => {
...
+                image: u.imageUrl,
+                createdAt: u.createdAt.getTime(),
```
Comment (flvndvd) [inline front/migrations/20240423_backfill_customerio.ts:38]: Those seems duplicated from the `toJSON` method.

File: front/pages/api/create-new-workspace.ts [hunk 33]
```diff
@@ -51,8 +52,20 @@ async function handler(
   }
 
   const workspace = await createWorkspace(session);
+  const userRes = await UserResource.fetchByModelId(user.id);
```
Comment (flvndvd) [inline front/pages/api/create-new-workspace.ts:55]: Why not going through the `Authenticator` here?

File: front/poke/temporal/activities.ts [hunk 34]
```diff
@@ -539,9 +533,9 @@ export async function deleteMembersActivity({
             `[Workspace delete] Deleting Membership ${membership.id} and user ${user.id}`
           );
           // Delete the user's files
-          await FileResource.deleteAllForUser(renderUserType(user), t);
+          await FileResource.deleteAllForUser(user.toJSON(), t);
           await membership.delete(auth, t);
-          await user.destroy({ transaction: t });
+          await user.delete();
```
Comment (flvndvd) [inline front/poke/temporal/activities.ts:538]: Why not passing the transaction here anymore?

File: front/pages/api/login.ts (hunk 21 above)
Comment (flvndvd) [reply front/pages/api/login.ts:447]: Having the resource in `lib` can be handy to avoid refetching it if actions need to be taken on the resource.

File: front/pages/no-workspace.tsx (hunk 23 above)
Comment (flvndvd) [reply front/pages/no-workspace.tsx:50]: For now, we should still return `UserType` from the Authenticator. Specifically here, we should never have Resources used in `pages` code. It seems way too easy to break something when trying to load server dependencies when running front-end code.

**Review by flvndvd — COMMENTED** [review]
🔥

File: front/poke/temporal/activities.ts (hunk 34 above)
Comment (albandum) [reply front/poke/temporal/activities.ts:538]: @spolu made a bunch of comments to remove transactions from the resource - so delete() doesn't have a transaction anymore

File: front/lib/iam/users.ts (hunk 15 above)
Comment (albandum) [reply front/lib/iam/users.ts:120]: Not sure what you mean by that - resource attributes are readonly so I can't just say `user.email = x`
You'd prefer doing a bunch of updates than building a newUser object? Or maybe I'm not seeing how you want to proceed

File: front/pages/api/create-new-workspace.ts (hunk 33 above)
Comment (albandum) [reply front/pages/api/create-new-workspace.ts:55]: The session contains a UserTypeWithWorkspace - because createAndLogMembership then uses the membership resource, we will need to get the UserResource at some point

File: front/pages/api/w/[wId]/members/[uId]/index.ts (hunk 22 above)
Comment (albandum) [reply front/pages/api/w/[wId]/members/[uId]/index.ts]: Renamed

File: front/pages/no-workspace.tsx (hunk 23 above)
Comment (albandum) [reply front/pages/no-workspace.tsx:50]: Not too sure what you're saying there - I shouldn't pulling the user resource?
Pb being that the getLatestMemberships expects a userResource. Should I change it so that it accepts both UserType and UserResource?

File: front/migrations/20240423_backfill_customerio.ts (hunk 20 above)
Comment (albandum) [reply front/migrations/20240423_backfill_customerio.ts:36]: Because ts is unhappy with it, as u is a model and not a resource
(Also that's just a backfill migration so it was a quickfix not trying to make it too clean)

File: front/lib/tracking/customerio/server.ts (hunk 30 above)
Comment (albandum) [reply front/lib/tracking/customerio/server.ts]: Done

File: front/lib/resources/user_resource.ts (hunk 31 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: Done

File: front/lib/resources/user_resource.ts (hunk 28 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: Done

File: front/lib/resources/user_resource.ts (hunk 27 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: Done - dunno what happened there 🤓

File: front/lib/resources/user_resource.ts (hunk 18 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: Done

File: front/lib/resources/user_resource.ts (hunk 17 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: done

File: front/lib/resources/user_resource.ts (hunk 16 above)
Comment (albandum) [reply front/lib/resources/user_resource.ts]: done (listby*)

Comment (albandum) [reply front/lib/resources/user_resource.ts]: Done

File: front/lib/api/user.ts (hunk 25 above)
Comment (albandum) [reply front/lib/api/user.ts]: Done

File: front/lib/iam/users.ts (hunk 15 above)
Comment (spolu) [reply front/lib/iam/users.ts:120]: More like cut a UserResource.updateInfo({ name, ..., createdAt: ...})

And then create that argument object and use that method

File: front/lib/resources/user_resource.ts (hunk 16 above)
Comment (spolu) [reply front/lib/resources/user_resource.ts]: listByModelIds as per @flvndvd comment on naming

File: front/pages/api/login.ts (hunk 21 above)
Comment (spolu) [reply front/pages/api/login.ts:446]: So this means that we likely want to go all the way to returning the resource in auth.user() no?

File: front/lib/iam/users.ts (hunk 15 above)
Comment (albandum) [reply front/lib/iam/users.ts]: Done

**Review by spolu — COMMENTED** [review]
Getting closer!

Principles to ensure:
- lib/api functions should not expose UserResource as return type if callers don't need to act on the resource. Keep UserType in that case
- lib/api functions should not accept USerResource as arguments if they don't need to interact with the resource and instead continue rely on UserType

As a consequence all of tracking/* should not be changed and receive UserType with .toJSON calls when going into them.

File: front/pages/api/login.ts [hunk 35]
```diff
@@ -400,7 +399,7 @@ async function handler(
 
       // Delete newly created user if SSO is mandatory.
       if (userCreated) {
-        await deleteUser(user);
+        await user.unsafeDelete();
```
Comment (albandum) [inline front/pages/api/login.ts:402]: I'm not super happy with that `unsafeDelete()` but I don't think I can get an Authenticator here as we don't seem to always know the workspaceId

File: front/lib/tracking/customerio/server.ts [hunk 36]
```diff
@@ -130,8 +131,18 @@ export class CustomerioServerSideTracking {
   }
 
   static async backfillUser({ user }: { user: UserType }) {
+    const userRes = await UserResource.fetchById(user.sId);
```
Comment (spolu) [inline front/lib/tracking/customerio/server.ts:134]: Should use `fetchNonNullableByModelId` since you have the ModelId in user.id ?

**Review by spolu — APPROVED** [review]
LGTM

Given the blast radius of this PR. Let's block on a LGTM from @flvndvd as well :pray: 

Can you also try to test most of the flows potentially impacted on front-edge (sign-in, membership edition, general use of the app) :pray:

**Comment by spolu** [comment]
Looks great!

File: front/lib/api/user.ts [hunk 37]
```diff
@@ -121,44 +101,24 @@ export async function setUserMetadata(
...
-  const u = await User.findOne({
-    where: {
-      id: user.id,
-    },
+export async function fetchRevokedWorkspace(
+  user: UserTypeWithWorkspaces
+): Promise<Workspace | null> {
+  // TODO(@fontanierh): this doesn't look very solid as it will start to behave
+  // weirdly if a user has multiple revoked memberships.
+  const userRes = await UserResource.fetchNonNullableByModelId(user.id);
```
Comment (flvndvd) [inline front/lib/api/user.ts:109]: `nit`
```suggestion
  const user = await UserResource.fetchNonNullableByModelId(user.id);
```

As the `Res` suffix is often use for our `Result`

File: front/lib/iam/users.ts [hunk 38]
```diff
@@ -95,46 +89,56 @@ export async function maybeUpdateFromExternalUser(
...
         );
-        user.firstName = firstName;
-        user.lastName = lastName;
+        updateArgs.firstName = firstName;
+        updateArgs.lastName = lastName;
       }
     }
 
-    await user.save();
+    await user.updateInfo(
```
Comment (flvndvd) [inline front/lib/iam/users.ts:125]: This is different from what we are currently doing, as `user.save()` was not doing a "saving" if the user has not been modified.

File: front/lib/resources/user_resource.ts [hunk 39]
```diff
@@ -0,0 +1,263 @@
...
+    const users = await User.findAll({
+      where: {
+        email,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchNonNullableByModelId(
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:74]: I'm not a big fan of this approach since it introduces more entry points and variations for the fetching method.

File: front/lib/api/user.ts [hunk 40]
```diff
@@ -121,44 +101,28 @@ export async function setUserMetadata(
...
 
-export async function unsafeGetUsersByModelId(
-  modelIds: number[]
-): Promise<UserType[]> {
-  if (modelIds.length === 0) {
-    return [];
+  if (!memberships.length) {
+    const message = "Unreachable: user has no memberships.";
+    logger.error({ userId: user.id, panic: true }, message);
+    throw new Error(message);
```
Comment (flvndvd) [inline front/lib/api/user.ts:122]: We should not throw here, but instead return a result.

File: front/lib/iam/users.ts [hunk 41]
```diff
@@ -95,46 +89,60 @@ export async function maybeUpdateFromExternalUser(
...
-        user.lastName = lastName;
+        updateArgs.firstName = firstName;
+        updateArgs.lastName = lastName || "";
       }
     }
 
-    await user.save();
+    if (Object.keys(updateArgs).length > 0) {
+      const needsUpdate = Object.entries(updateArgs).some(
+        ([key, value]) => user[key as keyof typeof user] !== value
```
Comment (flvndvd) [inline front/lib/iam/users.ts:123]: `nit` not a big fan, could be removed if user.updateInfo was taking an object 😉 

(can be fixed in a follow up PR)

File: front/lib/resources/user_resource.ts [hunk 42]
```diff
@@ -0,0 +1,250 @@
...
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  async unsafeDelete(
+    transaction?: Transaction
+  ): Promise<Result<undefined, Error>> {
+    try {
+      await this.model.destroy({
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:214]: `nit` at least we should ensure that the user does not have an active membership as this function should solely be used to delete users that did not pass the login.

(Can be fixed in a follow up PR).

File: front/lib/resources/user_resource.ts [hunk 43]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  getAttribute(attribute: string): any {
+    switch (attribute) {
+      case "fullName":
+        return [this.firstName, this.lastName].filter(Boolean).join(" ");
+      default:
+        return "";
+    }
+  }
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:234]: IHMO over-engineered, a simple, would have done the job
 ```suggestion
  get fullName(){
      return [this.firstName, this.lastName].filter(Boolean).join(" ");
  }
```

File: front/migrations/20240423_backfill_customerio.ts [hunk 44]
```diff
@@ -20,19 +20,41 @@ const backfillCustomerIo = async (execute: boolean) => {
     if (execute) {
       await Promise.all(
         c.map((u) =>
-          Promise.all([
-            CustomerioServerSideTracking.backfillUser({
-              user: u,
-            }).catch((err) => {
+          (async () => {
```
Comment (flvndvd) [inline front/migrations/20240423_backfill_customerio.ts:23]: FWIW, map does not support async function. I'd simply comment out this migration as it was a one time thing.

**Review by flvndvd — APPROVED** [review]
LGTM, left a few comments that can be addressed in a follow up PR as this one is already pretty big. Just fix the throw in this PR please 🙏

File: front/pages/no-workspace.tsx [hunk 45]
```diff
@@ -80,14 +80,16 @@ export const getServerSideProps = withDefaultUserAuthPaywallWhitelisted<{
...
 
-    if (!workspace) {
+    if (res.isErr()) {
       logger.error(
         { flow, userId: user.id, panic: true },
         "Unreachable: workspace not found."
       );
       throw new Error("Workspace not found.");
+    } else {
+      workspace = res.value;
```
Comment (flvndvd) [inline front/pages/no-workspace.tsx:92]: You don't need the else as you throw.

File: front/lib/resources/user_resource.ts (hunk 43 above)
Comment (flvndvd) [reply front/lib/resources/user_resource.ts]: Misunderstood then 🙈
